In [23]:
import pandas as pd

def save_excel_as_csv(excel_file_path, csv_file_path):

    data = pd.read_excel(excel_file_path)

    data.to_csv(csv_file_path, index=False)

excel_path = '/content/HighSchoolData.xlsx'
csv_path = '/content/data.csv'
save_excel_as_csv(excel_path, csv_path)


In [24]:

data = pd.read_csv(csv_path)


missing_values_per_column = data.isna().sum()
missing_values_per_row = data.isna().sum(axis=1)


missing_values_summary = {
    'Missing Values Per Column': missing_values_per_column,
    'Missing Values Per Row': missing_values_per_row
}

missing_values_summary


{'Missing Values Per Column': School Name                0
 City                       0
 Zip Code                   0
 Total Enrollments          1
 AP Classes                 3
 Dual Enrollment            3
 Offers Electives           2
 Offers Sports              1
 Mental Health Services     4
 Math Score                 0
 English Score              1
 National Rank             21
 AZ Rank                   13
 Student-Teacher Ratio      2
 Racial%-White              1
 Racial%-Black              1
 Racial%-Native             4
 Racial%-Hispanic           1
 Racial%-Asian              1
 Racial%-Other              1
 Lunch%-Free               10
 Lunch%-Reduced            11
 Lunch%-Paid               11
 dtype: int64,
 'Missing Values Per Row': 0      0
 1      2
 2      0
 3      2
 4      0
       ..
 219    0
 220    0
 221    0
 222    0
 223    0
 Length: 224, dtype: int64}

In [25]:
threshold = 0.5*len(data)
columns_to_drop = missing_values_per_column[missing_values_per_column > threshold].index
data_cleaned = data.drop(columns=columns_to_drop)
missing_values_per_column_cleaned = data_cleaned.isna().sum()
missing_values_per_row_cleaned = data_cleaned.isna().sum(axis=1)

updated_missing_values_summary = {
    'Missing Values Per Column After Initial Cleaning': missing_values_per_column_cleaned,
    'Missing Values Per Row After Initial Cleaning': missing_values_per_row_cleaned
}

updated_missing_values_summary

{'Missing Values Per Column After Initial Cleaning': School Name                0
 City                       0
 Zip Code                   0
 Total Enrollments          1
 AP Classes                 3
 Dual Enrollment            3
 Offers Electives           2
 Offers Sports              1
 Mental Health Services     4
 Math Score                 0
 English Score              1
 National Rank             21
 AZ Rank                   13
 Student-Teacher Ratio      2
 Racial%-White              1
 Racial%-Black              1
 Racial%-Native             4
 Racial%-Hispanic           1
 Racial%-Asian              1
 Racial%-Other              1
 Lunch%-Free               10
 Lunch%-Reduced            11
 Lunch%-Paid               11
 dtype: int64,
 'Missing Values Per Row After Initial Cleaning': 0      0
 1      2
 2      0
 3      2
 4      0
       ..
 219    0
 220    0
 221    0
 222    0
 223    0
 Length: 224, dtype: int64}

In [26]:
data_fully_cleaned = data_cleaned.dropna()
final_missing_values_per_column = data_fully_cleaned.isna().sum()
final_missing_values_per_row = data_fully_cleaned.isna().sum(axis=1)

final_missing_values_summary = {
    'Final Missing Values Per Column': final_missing_values_per_column,
    'Final Missing Values Per Row': final_missing_values_per_row
}
final_missing_values_summary, data_fully_cleaned.shape

({'Final Missing Values Per Column': School Name               0
  City                      0
  Zip Code                  0
  Total Enrollments         0
  AP Classes                0
  Dual Enrollment           0
  Offers Electives          0
  Offers Sports             0
  Mental Health Services    0
  Math Score                0
  English Score             0
  National Rank             0
  AZ Rank                   0
  Student-Teacher Ratio     0
  Racial%-White             0
  Racial%-Black             0
  Racial%-Native            0
  Racial%-Hispanic          0
  Racial%-Asian             0
  Racial%-Other             0
  Lunch%-Free               0
  Lunch%-Reduced            0
  Lunch%-Paid               0
  dtype: int64,
  'Final Missing Values Per Row': 0      0
  2      0
  4      0
  7      0
  8      0
        ..
  219    0
  220    0
  221    0
  222    0
  223    0
  Length: 180, dtype: int64},
 (180, 23))

Per Column: Different columns have varying numbers of missing values. For example, columns like 'School Name' and 'City' have no missing values, while others like 'National Rank' and 'AZ Rank' have large number of missing values.
Per Row: Some rows have no missing values at all, while others have up to 2 missing values.
To clean this data, I'll start by removing columns with the most missing values and then reassess the dataset. If necessary, we'll proceed with removing more columns or rows. Let's start with the column-wise cleaning.

Per Column: There are still columns with missing values, but none exceeds the 10% threshold we set.
Per Row: The number of missing values per row hasn't changed significantly, as we only removed columns in this step.
Now, rows with missing values are removed for further clean the dataset.

After the final cleaning step, the dataset is now fully cleaned of missing values:
Per Column: There are no missing values in any of the columns.
Per Row: There are no missing values in any of the rows.
The final shape of the cleaned dataset is 180 rows and 23 columns, indicating that some rows were removed to achieve this clean state.


**Q2**

In [27]:
duplicates = data_fully_cleaned.duplicated(subset=['School Name', 'City', 'Zip Code'], keep=False)
duplicate_records = data_fully_cleaned[duplicates]

duplicate_records_sorted = duplicate_records.sort_values(by=['School Name', 'City', 'Zip Code'])
duplicate_records_sorted

School Name          City  Zip Code  Total Enrollments  \
99           Apollo High School   Glendale,AZ     85302             2157.0   
121          Apollo High School   Glendale,AZ     85302             2157.0   
100          Cactus High School   Glendale,AZ     85306             1146.0   
122          Cactus High School   Glendale,AZ     85306             1146.0   
172     Campo Verde High School       Gilbert     85297             2091.0   
220     Campo Verde High School       Gilbert     85297             2091.0   
33         Chandler High School  Chandler, AZ     85225             3184.0   
118        Chandler High School  Chandler, AZ     85225             3184.0   
101   Copper Canyon High School   Glendale,AZ     85305             2195.0   
123   Copper Canyon High School   Glendale,AZ     85305             2195.0   
173    Desert Hills High School       Gilbert     85296              195.0   
221    Desert Hills High School       Gilbert     85296              195.0   
54   Fremont Junior High School      Mesa, AZ     85205             1012.0   
114  Fremont Junior High School      Mesa, AZ     85205             1012.0   
0          Hamilton High School  Chandler, AZ     85248             3926.0   
95         Hamilton High School  Chandler, AZ     85248             3926.0   
21         Highland High School   Gilbert, AZ     85234             3108.0   
26         Highland High School   Gilbert, AZ     85234             3108.0   
179        Maryvale High School       Phoenix     85033             2991.0   
222        Maryvale High School       Phoenix     85033             2991.0   
65             Mesa High School          Mesa     85204             3480.0   
213            Mesa High School          Mesa     85204             3480.0   
180      Metro Tech High School       Phoenix     85015             1716.0   
223      Metro Tech High School       Phoenix     85015             1716.0   
66    Mountain View High School          Mesa     85213             3343.0   
214   Mountain View High School          Mesa     85213             3343.0   
133    Red Mountain High School         Messa     85207             3370.0   
216    Red Mountain High School         Messa     85207             3370.0   
57          Saguaro High School    Scottsdale     85250             1292.0   
185         Saguaro High School    Scottsdale     85250             1292.0   

    AP Classes Dual Enrollment Offers Electives Offers Sports  \
99         yes             yes              yes           yes   
121        yes             yes              yes           yes   
100        yes             yes              yes           yes   
122        yes             yes              yes           yes   
172        Yes             Yes              Yes           Yes   
220        Yes             Yes              yes            no   
33         yes             yes              yes           yes   
118        Yes             Yes              Yes           Yes   
101        yes              no              yes           yes   
123        yes              no              yes           yes   
173        Yes              No              Yes           Yes   
221        Yes              No              yes           Yes   
54          no              No              yes           yes   
114         no              No              yes           yes   
0          yes             Yes              Yes           Yes   
95         yes              No              Yes           Yes   
21         YES             YES              YES           YES   
26         YES             YES              YES           YES   
179        Yes             Yes              yes           Yes   
222        Yes             Yes               no           yes   
65         yes             yes              yes           yes   
213        Yes             yes              yes           yes   
180        Yes             Yes              yes           Yes   
223        yes             yes           

### As we can see from the output cell above, There are 15 duplicate rows (30 in total) which needs to be removed.

In [28]:
data_no_duplicates = data_fully_cleaned.drop_duplicates(subset=['School Name', 'City', 'Zip Code'], keep='first')


post_duplicate_removal_count = data_no_duplicates.shape[0]
original_count = data_fully_cleaned.shape[0]
duplicates_removed = original_count - post_duplicate_removal_count

post_duplicate_removal_count, duplicates_removed

(165, 15)

In addressing the issue of duplicate records for the same school, the process was initiated by identifying duplicates based on a combination of 'School Name', 'City', and 'Zip Code'. Upon detecting duplicates, the typical approach would be to combine these records by averaging or summing relevant numeric values. However, in this specific case, **it appeared that the duplicate records were exact copies of each other rather than distinct entries that needed aggregation.**

Therefore, we retained only the first occurrence of each duplicate set and discard the subsequent duplicates. This method ensures no loss of information or distortion of values. It's important to note that if the duplicates had contained varying data points, a complex method involving averaging or summing of specific columns would have been necessary.

The final dataset post this duplicate removal contains distinct entries for each school without any redundancies.

**Q3**

In [29]:
def clean_numeric_columns(df):
    for column in df.columns:
        if df[column].dtype == 'object':
            df[column] = pd.to_numeric(df[column].replace('[\$,%,]', '', regex=True), errors='ignore')
    return df
data_transformed = clean_numeric_columns(data_no_duplicates)
data_transformed

<ipython-input-29-2f8be05995c9>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[column] = pd.to_numeric(df[column].replace('[\$,%,]', '', regex=True), errors='ignore')


School Name         City  Zip Code  \
0                     Hamilton High School  Chandler AZ     85248   
2                     Mesquite High School   Gilbert AZ     85233   
4                        Perry High School   Gilbert AZ     85297   
7                    Camelback High School   Phoenix AZ     85016   
8    Career Success High School - Glendale   Phoenix AZ     85017   
..                                     ...          ...       ...   
212   Franklin Police and Fire High School      Phoenix     85007   
215              Poston Junior High School        Messa     85213   
217             Mountain Ridge High School     Glendale     85310   
218        Sandra Day O'Connor High School     Glendale     85310   
219                Deer Valley High School     Glendale     85308   

     Total Enrollments AP Classes Dual Enrollment Offers Electives  \
0               3926.0        yes             Yes              Yes   
2               2045.0        Yes             Yes              Yes   
4               2045.0        Yes             Yes              Yes   
7               2152.0        yes             yes              yes   
8                 53.0        yes             yes              yes   
..                 ...        ...             ...              ...   
212              287.0       Yes             Yes               Yes   
215             1017.0        Yes              No              Yes   
217             2685.0        yes             yes              yes   
218             2534.0        yes             yes              yes   
219             1420.0        YES             YES              YES   

    Offers Sports Mental Health Services  Math Score  ...  \
0             Yes                     No        52.0  ...   
2             Yes                    Yes        80.0  ...   
4             Yes                    Yes        80.0  ...   
7             yes                    yes        80.0  ...   
8              no                    yes        75.0  ...   
..            ...                    ...         ...  ...   
212            No                     No        19.0  ...   
215            No                     No        24.0  ...   
217           yes                    yes        51.0  ...   
218           yes                    yes        70.0  ...   
219           YES                    YES        58.0  ...   

     Student-Teacher Ratio Racial%-White Racial%-Black Racial%-Native  \
0                     21:1          44.4           7.3            2.1   
2                     23:1          45.9           4.8            2.1   
4                     20:1          67.1           3.4            0.6   
7                     18:1           8.0           7.0            1.0   
8                     17:1           8.0          11.0            0.0   
..                     ...           ...           ...            ...   
212                   10:1           3.1           7.0           17.0   
215                   18:1          43.0           5.0            3.0   
217                   25:1          69.0           2.0            1.0   
218                   23:1          74.0           2.0            1.0   
219                   15:1          55.2           5.9            1.2   

     Racial%-Hispanic  Racial%-Asian  Racial%-Other  Lunch%-Free  \
0                21.5           19.0            5.7            8   
2                38.1            2.8            6.3           12   
4                17.5            8.1            3.3           20   
7                77.0            1.0            6.0           82   
8                77.0            0.0            4.0           61   
..                ...            ...            ...          ...   
212              93.4            1.0            0.0           73   
215              47.0            1.0            1.0           58   
217              13.0           10.0            5.0            4   
218              14.0            5.0            4.0            5   
219

The final dataset, after cleaning and transformation, consists of 165 rows and 23 columns.

Initial Assessment:

Calculate a summary of missing values per column and row to quantify the extent of missing data.
Establish a threshold for acceptable levels of missing data. In this case, columns with more than 10% missing values were targeted for removal.
After reassessing the dataset, proceed to eliminate rows with any missing values. This ensures a dataset with complete records.
Handling Duplicate Records:

Remove duplicate records while keeping the first occurrence of each duplicate set. This process ensures each school is represented once, maintaining data integrity.

In [30]:
binary_columns = ['AP Classes', 'Dual Enrollment', 'Offers Electives', 'Offers Sports', 'Mental Health Services']
data_no_duplicates[binary_columns] = data_no_duplicates[binary_columns].apply(lambda x: x.str.lower().map({'yes': 1, 'no': 0}))

<ipython-input-30-ad5e2079e59c>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_no_duplicates[binary_columns] = data_no_duplicates[binary_columns].apply(lambda x: x.str.lower().map({'yes': 1, 'no': 0}))


In [31]:
data_no_duplicates

School Name         City  Zip Code  \
0                     Hamilton High School  Chandler AZ     85248   
2                     Mesquite High School   Gilbert AZ     85233   
4                        Perry High School   Gilbert AZ     85297   
7                    Camelback High School   Phoenix AZ     85016   
8    Career Success High School - Glendale   Phoenix AZ     85017   
..                                     ...          ...       ...   
212   Franklin Police and Fire High School      Phoenix     85007   
215              Poston Junior High School        Messa     85213   
217             Mountain Ridge High School     Glendale     85310   
218        Sandra Day O'Connor High School     Glendale     85310   
219                Deer Valley High School     Glendale     85308   

     Total Enrollments  AP Classes  Dual Enrollment  Offers Electives  \
0               3926.0         1.0              1.0               1.0   
2               2045.0         1.0              1.0               1.0   
4               2045.0         1.0              1.0               1.0   
7               2152.0         1.0              1.0               1.0   
8                 53.0         1.0              1.0               1.0   
..                 ...         ...              ...               ...   
212              287.0         NaN              NaN               1.0   
215             1017.0         1.0              0.0               1.0   
217             2685.0         1.0              1.0               1.0   
218             2534.0         1.0              1.0               1.0   
219             1420.0         1.0              1.0               1.0   

     Offers Sports  Mental Health Services  Math Score  ...  \
0              1.0                     0.0        52.0  ...   
2              1.0                     1.0        80.0  ...   
4              1.0                     1.0        80.0  ...   
7              1.0                     1.0        80.0  ...   
8              0.0                     1.0        75.0  ...   
..             ...                     ...         ...  ...   
212            0.0                     0.0        19.0  ...   
215            0.0                     0.0        24.0  ...   
217            1.0                     1.0        51.0  ...   
218            1.0                     1.0        70.0  ...   
219            1.0                     1.0        58.0  ...   

     Student-Teacher Ratio Racial%-White Racial%-Black Racial%-Native  \
0                     21:1          44.4           7.3            2.1   
2                     23:1          45.9           4.8            2.1   
4                     20:1          67.1           3.4            0.6   
7                     18:1           8.0           7.0            1.0   
8                     17:1           8.0          11.0            0.0   
..                     ...           ...           ...            ...   
212                   10:1           3.1           7.0           17.0   
215                   18:1          43.0           5.0            3.0   
217                   25:1          69.0           2.0            1.0   
218                   23:1          74.0           2.0            1.0   
219                   15:1          55.2           5.9            1.2   

     Racial%-Hispanic  Racial%-Asian  Racial%-Other  Lunch%-Free  \
0                21.5           19.0            5.7            8   
2                38.1            2.8            6.3           12   
4                17.5            8.1            3.3           20   
7                77.0            1.0            6.0           82   
8                77.0            0.0            4.0           61   
..                ...            ...            ...          ...   
212              93.4            1.0            0.0           73   
215              47.0            1.0            1.0           58   
217              13.0           10.0            5.0            4   
218        

In [32]:
data_no_duplicates.dropna(inplace=True)

<ipython-input-32-599789a9d4c7>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_no_duplicates.dropna(inplace=True)


In [33]:
df = data_no_duplicates.drop(columns=['City'])
df

School Name  Zip Code  Total Enrollments  \
0                     Hamilton High School     85248             3926.0   
2                     Mesquite High School     85233             2045.0   
4                        Perry High School     85297             2045.0   
7                    Camelback High School     85016             2152.0   
8    Career Success High School - Glendale     85017               53.0   
..                                     ...       ...                ...   
210             Stapley Junior High School     85213              956.0   
215              Poston Junior High School     85213             1017.0   
217             Mountain Ridge High School     85310             2685.0   
218        Sandra Day O'Connor High School     85310             2534.0   
219                Deer Valley High School     85308             1420.0   

     AP Classes  Dual Enrollment  Offers Electives  Offers Sports  \
0           1.0              1.0               1.0            1.0   
2           1.0              1.0               1.0            1.0   
4           1.0              1.0               1.0            1.0   
7           1.0              1.0               1.0            1.0   
8           1.0              1.0               1.0            0.0   
..          ...              ...               ...            ...   
210         1.0              1.0               1.0            1.0   
215         1.0              0.0               1.0            0.0   
217         1.0              1.0               1.0            1.0   
218         1.0              1.0               1.0            1.0   
219         1.0              1.0               1.0            1.0   

     Mental Health Services  Math Score  English Score  ...  \
0                       0.0        52.0          68.96  ...   
2                       1.0        80.0          75.00  ...   
4                       1.0        80.0          75.00  ...   
7                       1.0        80.0          75.00  ...   
8                       1.0        75.0          70.00  ...   
..                      ...         ...            ...  ...   
210                     0.0         5.0          15.00  ...   
215                     0.0        24.0          33.00  ...   
217                     1.0        51.0          55.00  ...   
218                     1.0        70.0          69.00  ...   
219                     1.0        58.0          51.00  ...   

    Student-Teacher Ratio Racial%-White Racial%-Black  Racial%-Native  \
0                    21:1          44.4           7.3             2.1   
2                    23:1          45.9           4.8             2.1   
4                    20:1          67.1           3.4             0.6   
7                    18:1           8.0           7.0             1.0   
8                    17:1           8.0          11.0             0.0   
..                    ...           ...           ...             ...   
210                  20:1          70.0           1.0             8.0   
215                  18:1          43.0           5.0             3.0   
217                  25:1          69.0           2.0             1.0   
218                  23:1          74.0           2.0             1.0   
219                  15:1          55.2           5.9             1.2   

     Racial%-Hispanic  Racial%-Asian  Racial%-Other  Lunch%-Free  \
0                21.5           19.0            5.7            8   
2                38.1            2.8            6.3           12   
4                17.5            8.1            3.3           20   
7                77.0            1.0            6.0           82   
8                77.0            0.0            4.0           61   
..                ...            ...            ...          ...   
210              18.0            2.0            9.0           25   
215              47.0            1.0            1.0           58   
217              13.0           10.0            5.0    

In [34]:
x=pd.get_dummies(data_no_duplicates['City'],dtype = 'int')
data_new = pd.concat([df,x],axis=1)

data_new

School Name  Zip Code  Total Enrollments  \
0                     Hamilton High School     85248             3926.0   
2                     Mesquite High School     85233             2045.0   
4                        Perry High School     85297             2045.0   
7                    Camelback High School     85016             2152.0   
8    Career Success High School - Glendale     85017               53.0   
..                                     ...       ...                ...   
210             Stapley Junior High School     85213              956.0   
215              Poston Junior High School     85213             1017.0   
217             Mountain Ridge High School     85310             2685.0   
218        Sandra Day O'Connor High School     85310             2534.0   
219                Deer Valley High School     85308             1420.0   

     AP Classes  Dual Enrollment  Offers Electives  Offers Sports  \
0           1.0              1.0               1.0            1.0   
2           1.0              1.0               1.0            1.0   
4           1.0              1.0               1.0            1.0   
7           1.0              1.0               1.0            1.0   
8           1.0              1.0               1.0            0.0   
..          ...              ...               ...            ...   
210         1.0              1.0               1.0            1.0   
215         1.0              0.0               1.0            0.0   
217         1.0              1.0               1.0            1.0   
218         1.0              1.0               1.0            1.0   
219         1.0              1.0               1.0            1.0   

     Mental Health Services  Math Score  English Score  ... Queen Creek AZ  \
0                       0.0        52.0          68.96  ...              0   
2                       1.0        80.0          75.00  ...              0   
4                       1.0        80.0          75.00  ...              0   
7                       1.0        80.0          75.00  ...              0   
8                       1.0        75.0          70.00  ...              0   
..                      ...         ...            ...  ...            ...   
210                     0.0         5.0          15.00  ...              0   
215                     0.0        24.0          33.00  ...              0   
217                     1.0        51.0          55.00  ...              0   
218                     1.0        70.0          69.00  ...              0   
219                     1.0        58.0          51.00  ...              0   

    Scottsdale Scottsdale AZ  Surprise AZ  Tempe  Tempe AZ  TempeAZ  Tolleson  \
0            0             0            0      0         0        0         0   
2            0             0            0      0         0        0         0   
4            0             0            0      0         0        0         0   
7            0             0            0      0         0        0         0   
8            0             0            0      0         0        0         0   
..         ...           ...          ...    ...       ...      ...       ...   
210          0             0            0      0         0        0         0   
215          0             0            0      0         0        0         0   
217          0             0            0      0         0        0         0   
218          0             0            0      0         0        0         0   
219          0             0            0      0         0        0         0   

     TonopahAz WikenburgAz  
0            0           0  
2            0           0  
4            0           0  
7            0           0  
8            0           0  
..         ...         ...  
210          0           0  
215          0           0  
217          0           0  
218          0           0  
219          0           0  

[159 rows x 58 columns]

For the data cleaning and transformation process, it's important to consider the attribute types of the data. These types can include nominal, ordinal, interval, and ratio.

Nominal Data: Transformed city attribute using one hot encoding.

Ratio Data: In this dataset include 'Total Enrollments', 'Math Score', 'English Score'. The transformation for these data types involved cleaning non-numeric characters (like %, $) and converting them to a suitable numeric format for analysis.

Special Cases: Attributes like 'Student-Teacher Ratio' could be considered ratio data, but they were in a composite format (e.g., "18:1").

During the cleaning process, handling missing and duplicate data was crucial. Missing values were identified and rows/columns with excessive missing data were removed. Duplicate records were detected based on 'School Name', 'City', and 'Zip Code', and redundant entries were eliminated, ensuring each school was uniquely represented.

The columns having Yes/No as their column value, we convarted them to binary data (0/1). (I.E. 'AP Classes', 'Dual Enrollment', 'Offers Electives', 'Offers Sports', 'Mental Health Services')


Final dataset with all cleaning and transdormation contains 159 rows and 58 columns.



